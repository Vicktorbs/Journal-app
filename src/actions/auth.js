import { firebase, googleAuthProvider } from "../firebase/firebase_config";
import { types } from "../types/types"

export const startLoginWithEmailPassword = (email, password) => {
    return (dispatch) => {
        setTimeout(() => {
            dispatch(login(12345, 'Juan'))
        }, 3500);
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName))
        })
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}