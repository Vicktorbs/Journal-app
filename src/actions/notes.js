import Swal from "sweetalert2";
import { db } from "../firebase/firebase_config";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async(dispatch, getState) => {
        const uid = getState().auth.uid;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const documentRef = await db.collection(`${ uid }/journal/notes`).add(newNote);
        // console.log(documentRef);
        dispatch(activeNote(documentRef.id, newNote));
        // funcionalidad personalizada, checar
        dispatch(startLoadingNotes(uid));
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {
        const uid = getState().auth.uid;

        if (!note.url) {
            delete note.url
        }

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        await db.doc(`${ uid }/journal/notes/${ note.id }`).update(noteToFirestore);

        dispatch(refresNote(note.id, noteToFirestore));
        Swal.fire('Save', note.title, 'success')
    }
}

export const refresNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})