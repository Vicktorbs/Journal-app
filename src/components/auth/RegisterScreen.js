import React from 'react';
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';
import validator from 'validator'
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
    
    const dispatch = useDispatch();
    const { msgError } = useSelector(state => state.ui);

    const [values, handleInputChange] = useForm({
        name: 'Brenda',
        email: 'yoana@gmail.com',
        password: '123456',
        passwordConfirm: '123456'
    });

    const { name, email, password, passwordConfirm } = values;

    const handleSubmit = (e) => {
        e.preventDefault();
        // console.log(name, email, password, passwordConfirm);
        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name))
        }
    }

    const isFormValid = () => {
        
        if (name.trim().length === 0) {
            dispatch(setError('Name required'));
            // console.log('Name required');
            return false
        } else if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            // console.log('Email is not valid');
            return false
        } else if (password !== passwordConfirm || password.length < 5) {
            dispatch(setError('Password shold have at least 5 characters and be eaquals'));
            // console.log('Password shold have at least 5 characters and be eaquals');
            return false
        }

        dispatch(removeError());

        return true
    }

    return (
        <>
            <h2 className="auth_title">Register</h2>

            <form onSubmit={ handleSubmit } className="animate__animated animate__fadeIn animate__faster" >
                {
                    msgError && (
                    <div className="auth__alert-error">
                        {msgError}
                    </div>
                    )
                }
                <input 
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={ name }
                    onChange={ handleInputChange }
                />
                <input 
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />
                <input 
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={ password }
                    onChange={ handleInputChange }
                />
                <input 
                    type="password"
                    placeholder="Confirm password"
                    name="passwordConfirm"
                    className="auth__input"
                    value={ passwordConfirm }
                    onChange={ handleInputChange }
                />
                <button type="submit" className="btn btn-primary btn-block mb-5">
                    Register
                </button>

                <Link to="/auth/login" className="link">
                    Already registered?
                </Link>
            </form>
        </>
    )
}
