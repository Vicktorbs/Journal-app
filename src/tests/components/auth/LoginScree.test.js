import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store'; 
import thunk from 'redux-thunk';
import { startGoogleLogin, startLoginWithEmailPassword } from '../../../actions/auth';
import { LoginScree } from '../../../components/auth/LoginScree';

jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn(),
    startLoginWithEmailPassword: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initialState);
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <LoginScree />
        </MemoryRouter>
    </Provider>
)

describe('Test on component <LoginScree />', () => {

    beforeEach(() => {
        store = mockStore(initialState);
        jest.clearAllMocks();
    })

    test('should show itself correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    test('should lauch the action startGoogleLogin', () => {
        wrapper.find('.google-btn').prop('onClick')();

        expect(startGoogleLogin).toHaveBeenCalled()
    })
    
    test('should lauch startLoginWithEmailPassword witht the defaluy values', () => {
        wrapper.find('form').prop('onSubmit')({ preventDefault(){} });
        const values = {
            email: '',
            password: ''
        }
        expect(startLoginWithEmailPassword).toHaveBeenCalledWith(values.email, values.password)
    })
    
})
