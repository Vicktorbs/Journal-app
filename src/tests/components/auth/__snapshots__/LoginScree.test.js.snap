// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on component <LoginScree /> should show itself correctly 1`] = `
Array [
  <h2
    className="auth_title"
  >
    LoginScree
  </h2>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <div
      className="auth__socila-networks"
    >
      <p>
        Login with social networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    >
      Create new account
    </a>
  </form>,
]
`;
